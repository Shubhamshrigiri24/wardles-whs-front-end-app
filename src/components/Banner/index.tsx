import React, { FC } from "react";
import { Typography } from "@material-ui/core";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import { makeStyles, Theme } from "@material-ui/core/styles";
import { createPortal } from "react-dom";
import { usePortal } from "../../components";

import { BannerContainer } from "./components/BannerContainer";

const useStyles = makeStyles((theme: Theme) => ({
  banner: {
    background: theme.palette.grey[500],
    border: `1px solid ${theme.palette.divider}`,
    marginBottom: theme.spacing(3),
    display: "flex",
    flexDirection: "row",
  },
  iconContainer: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    margin: theme.spacing(0, 4),
  },
  textContainer: {
    margin: theme.spacing(3, 3, 3, 0),
  },
  textColor: {
    color: theme.palette.grey["A700"],
  },
}));

export interface BannerProps {
  variant?: "warning";
}

const Banner: FC<BannerProps> = (props) => {
  const { variant = "warning", children } = props;
  const classes = useStyles();

  return (
    <div className={classes.banner}>
      <div className={classes.iconContainer}>
        {variant === "warning" && (
          <InfoOutlinedIcon className={classes.textColor} />
        )}
      </div>
      <div className={classes.textContainer}>
        <Typography
          data-testid={"banner/text-content"}
          className={classes.textColor}
        >
          {children}
        </Typography>
      </div>
    </div>
  );
};

/**
 * Banner component that requires the BannerContainer to have been used somewhere
 * within the application as this relies on react portals.
 * Not using the BannerContainer would result in the banner being appended to
 * the`document.body` which would be undesired.
 * @example
 * // Banner container location
 * const BannerContainerLocation: FC = () => (
 *  <BannerContainer />
 * );
 * // Page where the banner is in regards to
 * const Page: FC = () => (
 *  <>
 *    <Banner>Banner Text</Banner>
 *    ...
 *  </>
 * );
 */
export const BannerPortals: React.FC<BannerProps> = (props) => {
  const banner = usePortal("banner-portal");
  const bannerMobile = usePortal("banner-portal-mobile");

  return (
    <>
      {createPortal(<Banner {...props} />, banner)}
      {createPortal(<Banner {...props} />, bannerMobile)}
    </>
  );
};

export default BannerPortals;
export { BannerContainer };
