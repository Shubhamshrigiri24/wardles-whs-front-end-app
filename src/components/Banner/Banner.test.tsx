import React from "react";
import { render } from "@testing-library/react";

import Banner from "./";

describe("<Banner />", () => {
  const text =
    "Here is a piece of text to demonstrate what a banner would look like on the page";

  it("should match the snapshot", () => {
    const { baseElement } = render(<Banner>{text}</Banner>);

    expect(baseElement).toMatchSnapshot();
  });

  it("should have content that matches the text props", () => {
    const { getAllByTestId } = render(<Banner>{text}</Banner>);

    getAllByTestId("banner/text-content").forEach(({ textContent }) => {
      expect(textContent).toBe(text);
    });
  });
});
