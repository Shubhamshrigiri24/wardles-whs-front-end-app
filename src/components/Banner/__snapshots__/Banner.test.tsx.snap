// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should match the snapshot 1`] = `
<body>
  <div />
  <div
    id="banner-portal"
  >
    <div
      id="banner-portal-inner"
    >
      <div
        class="makeStyles-banner-1"
      >
        <div
          class="makeStyles-iconContainer-2"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-textColor-4"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="makeStyles-textContainer-3"
        >
          <p
            class="MuiTypography-root makeStyles-textColor-4 MuiTypography-body1"
            data-testid="banner/text-content"
          >
            Here is a piece of text to demonstrate what a banner would look like on the page
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    id="banner-portal-mobile"
  >
    <div
      id="banner-portal-mobile-inner"
    >
      <div
        class="makeStyles-banner-1"
      >
        <div
          class="makeStyles-iconContainer-2"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-textColor-4"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="makeStyles-textContainer-3"
        >
          <p
            class="MuiTypography-root makeStyles-textColor-4 MuiTypography-body1"
            data-testid="banner/text-content"
          >
            Here is a piece of text to demonstrate what a banner would look like on the page
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
